pyop2/host.py

```diff
@@ -187,6 +222,47 @@ def c_kernel_arg(self, count, i=0, j=0, shape=(0,), is_top=False, layers=1):
             return "%(name)s + i * %(dim)s" % {'name': self.c_arg_name(i),
                                                'dim': self.data[i].cdim}
 
+    def ast_kernel_arg(self, count, i=0, j=0, shape=(0,), is_top=False, layers=1):
+        if self._uses_itspace:
+            if self._is_mat:
+                if self.data[i, j]._is_vector_field:
+                    return ast.Symbol(self.c_kernel_arg_name(i, j))
+                elif self.data[i, j]._is_scalar_field:
+                    flat_block = "(%(t)s (*)[%(dim)d])&%(name)s" % \
+                        {'t': self.ctype,
+                         'dim': shape[0],
+                         'name': self.c_kernel_arg_name(i, j)}
+                    return ast.FlatBlock(flat_block)
```

Comment:
We want to avoid FlatBlocks always, when possible - and this is one of those cases